### PR TITLE
Fix #1852

### LINKED ISSUE #1852: Lunch constraints should allow having an empty timeblock instead of signing up for a lunch
This would make things a lot simpler and prevent some of the common lunch constraints bugs and UI issues, and I think it actually wouldn't be hard to implement -- we'd just need to make the lunch constraint also have an extra AND for "has a class during all blocks of lunch".

### PR DESCRIPTION
**Pull request by abbysallord**
## Description

This PR changes the lunch constraint logic so students are only required to select a lunch period if they have classes on BOTH sides of lunch (morning AND afternoon), not just either one. This allows students who only have morning or afternoon classes to have an empty timeblock instead of being forced to sign up for lunch.

## Related Issue

Closes #1852 

## Type of Change

- [ ] Bug fix
- [x] New feature
- [ ] Breaking change
- [ ] Documentation update
- [ ] Refactor / cleanup
- [ ] CI/CD or build change

## Testing

- [x] Existing tests pass
- [ ] New tests added (if applicable)
- [x] Manually verified

## AI Disclosure

used antigravity for code assist and verified the code. I bear full responsibility of the code written by me.

## Checklist

- [x] Self-reviewed the code
- [ ] Updated documentation (if needed)
- [x] No new warnings or errors introduced

### CODE REVIEW

**Comment by abbysallord** [comment]
I have made an attempt to fix this issue, please review it

File: esp/esp/program/controllers/lunch_constraints.py [hunk 1]
```diff
@@ -235,15 +235,15 @@ def generate_constraint(self, day):
...
+            #   classes both before AND after lunch (not either one)
             morning_tests = []
             for timeslot in self.days[day]['before']:
                 new_test = ScheduleTestOccupied()
                 new_test.timeblock_id = timeslot.id
                 new_test.exp_id = exp_check.id
                 morning_tests.append(new_test)
                 seq_id += 1
-            self.apply_binary_op_to_list(exp_check, 'OR', '0', morning_tests)
+            self.apply_binary_op_to_list(exp_check, 'AND', '0', morning_tests)
```
Comment (Dhirenderchoudhary) [inline esp/esp/program/controllers/lunch_constraints.py]: [morning_tests](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html) and [afternoon_tests](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html) are now combined with AND. That makes the condition effectively require all morning and all afternoon slots to be occupied, which is stricter than intended and can evaluate false in normal schedules (especially single-slot edge cases with current reducer identity).

I think we should keep:

OR within morning tests
OR within afternoon tests
final AND between morning and afternoon groups

Comment (abbysallord) [reply esp/esp/program/controllers/lunch_constraints.py]: Got it. Will fix it right now.

**Comment by abbysallord** [comment]
I have fixed the issue @Dhirenderchoudhary . Please review the code for any further changes.

File: esp/esp/program/modules/handlers/teacherclassregmodule.py [hunk 2]
```diff
@@ -836,6 +836,37 @@ def editclass(self, request, tl, one, two, module, extra, prog):
...
+        
+        cls = classes[0]
+        saved = False
+
+        if request.method == 'POST':
+            for sec in cls.sections.all():
+                key = 'capacity_%d' % sec.id
+                if key in request.POST:
+                    try:
+                        new_cap = int(request.POST[key])
```
Comment (moksha-hub) [inline esp/esp/program/modules/handlers/teacherclassregmodule.py:860]: The new handler accepts any non-negative integer here, including values below the section's current enrollment. That lets a teacher save `max_class_capacity` smaller than `sec.num_students()`, which leaves the section in an impossible over-capacity state even though the template explicitly tells them not to do that. Can we reject values below the current enrollment count and surface a form error instead of silently saving them?

Comment (abbysallord) [reply esp/esp/program/modules/handlers/teacherclassregmodule.py:860]: Makes sense. I will fix it

**Comment by abbysallord** [comment]
I will get this fixed asap.

**Comment by abbysallord** [comment]
I have fixed it. Please review the changes @moksha-hub

**Comment by abbysallord** [comment]
I will get these issues asap. Give me sometime.

**Comment by abbysallord** [comment]
Please review the changes and let me know if any changes are required

**Review by Oval17 — CHANGES_REQUESTED** [review]
Hi @abbysallord 
There are a lot of repeated logic https://github.com/learning-unlimited/ESP-Website/pull/5589
Would suggest closing any one pr in favour of other and combining unique changes
